zio-http/src/main/scala/zhttp/service/server/ServerRequestHandler.scala

```diff
@@ -27,29 +27,36 @@ final case class ServerRequestHandler[R, E: SilentResponse](
   private def webSocketUpgrade(
     ctx: JChannelHandlerContext,
     jReq: JFullHttpRequest,
-    res: Response.SocketResponse[R, E],
+    res: Response.SocketResponse[R, Throwable],
   ): Unit = {
-    val hh = new JWebSocketServerHandshakerFactory(jReq.uri(), res.subProtocol.orNull, false).newHandshaker(jReq)
+    val settings = res.ss.settings
+    val hh       = new JWebSocketServerHandshakerFactory(jReq.uri(), settings.subProtocol.orNull, false).newHandshaker(jReq)
     if (hh == null) {
       JWebSocketServerHandshakerFactory.sendUnsupportedVersionResponse(ctx.channel, ctx.channel().voidPromise())
     } else {
       val pl = ctx.channel().pipeline()
-      pl.addLast(WEB_SOCKET_HANDLER, ServerSocketHandler(zExec, res.socket))
+      pl.addLast(WEB_SOCKET_HANDLER, ServerSocketHandler(zExec, settings))
       try {
         // handshake can throw
         hh.handshake(ctx.channel(), jReq).addListener { (future: JChannelFuture) =>
+          // FAILURE CONDITION
           if (!future.isSuccess) {
             pl.remove(WEB_SOCKET_HANDLER)
             ctx.fireExceptionCaught(future.cause)
+            zExec.unsafeExecute_(ctx)(settings.onError(future.cause()))
             ()
           } else {
+            // SUCCESS
             pl.remove(HTTP_REQUEST_HANDLER)
+            // TODO: handle channel closure on failure
+            zExec.unsafeExecute_(ctx)(settings.onOpen(ctx.channel().remoteAddress()))
             ()
           }
         }
       } catch {
-        case _: JWebSocketHandshakeException =>
+        case cause: JWebSocketHandshakeException =>
           JWebSocketServerHandshakerFactory.sendUnsupportedVersionResponse(ctx.channel, ctx.channel().voidPromise())
+          zExec.unsafeExecute_(ctx)(settings.onError(cause))
```

Comment:
Passing control to the server isn't required in this case. The handlers should be called only once we have upgrade to websocket.

Reply:
- [ ] We can use `ctx.fireChannelException` for all **unhandled** exceptions.

Reply:
Check this — https://github.com/dream11/zio-http/pull/123/files#r608738940